### ISSUE TITLE
precompiling cfunction calls

### ISSUE THREAD

**Issue by stevengj** [JuliaLang/julia#12256]
I noticed in stevengj/PyCall.jl#167 that, even after running `Base.compile`, some `cfunction` calls in my `__init__` function were taking a significant amount of time (0.8 seconds for 6 functions).

I could work around this by adding some explicit `precompile` lines in order to ensure that these were compiled, but this raises three problems:

* Calls to `cfunction` in `__init__` should really cause the corresponding functions to be precompiled automatically.

* Even after precompiling, `cfunction` was quite expensive: 0.25 seconds for 6 functions.

* Calling `precompile` causes the functions to be compiled at the wrong time — *before* `__init__` is called rather than *after*.  Since my `__init__` initializes some `global const` variables that are referenced in the precompiled functions, this means that type inference does not work properly when precompiling (since Julia does not know the types of those globals yet).

cc: @vtjnash

**Comment by vtjnash** [comment]
eventually, we need to disable the speed boost provided by calling `global const` from `__init__` since it causes potentially serious bugs in type inference / constant propagation: (https://github.com/JuliaLang/julia/commit/e2d842adeadfb6faca33260fd4190d89066bcafc#diff-c3408de56388e517a8975797c90d9448R385). although we'll likely need to implement the some alternative around the same time

**Comment by JeffBezanson** [comment]
One thing that would be neat, and should be entirely safe, is to automatically do `precompile(__init__, ())` as part of `compile`.

**Comment by vtjnash** [comment]
i'm pretty sure you already thought of that before, and implemented it (https://github.com/JuliaLang/julia/blob/8294d4f9922cadc3ffae7882d13a81fc82f515b0/src/toplevel.c#L85)

**Comment by stevengj** [comment]
Yes, `precompile(__init__, ())` made no difference, so it seems like `jl_get_specialization` is not specializing the function in the `cfunction` calls within `__init__` (or any other function) unless you `precompile` them explicitly..

**Comment by stevengj** [comment]
Unfortunately, this doesn't seem to be completely fixed.   When I remove the [manual precompilation from PyCall/src/PyCall.jl](https://github.com/stevengj/PyCall.jl/blob/59c2b950c2c209c291b379b26e64d3c157818a0c/src/PyCall.jl#L499-L513), the load time goes from 1.25s to 1.68s on my machine.

**Comment by vtjnash** [comment]
we pre-infer `__init__`, but i'm not sure if we pre-compile it

**Comment by stevengj** [comment]
@vtjnash, adding `precompile(__init__, ())` makes no difference, so that's not the problem.
